### PR TITLE
DHP-1085 Add phase enum field to Assessment

### PR DESCRIPTION
**Pull request by nategbrown9**
Still need to:

- [ ] Update Bridge Java Client
- [ ] Add/Update integration tests
- [ ] Update documentation

### CODE REVIEW

File: src/test/java/org/sagebionetworks/bridge/services/AssessmentServiceTest.java [hunk 1]
```diff
@@ -520,7 +522,82 @@ public void updateAssessmentCannotUpdatedDeleted() {
...
+
+        when(mockDao.getAssessment(TEST_APP_ID, TEST_OWNER_ID, assessment.getGuid()))
+                .thenReturn(Optional.of(existing));
+
+        service.updateAssessment(TEST_APP_ID, TEST_OWNER_ID, assessment);
+        assertEquals(assessment.getPhase(), AssessmentPhase.PUBLISHED);
+    }
+
+    @Test(expectedExceptions = BadRequestException.class)
+    public void updateAssessmentPhaseFromPublished() {
```
Comment (Erin-Mounts) [inline src/test/java/org/sagebionetworks/bridge/services/AssessmentServiceTest.java]: Should there also be tests for draft -> published and published -> review?

File: src/test/java/org/sagebionetworks/bridge/services/StudyServiceTest.java [hunk 2]
```diff
@@ -771,7 +779,14 @@ public void transitionToRecruitment() {
         study.setScheduleGuid(SCHEDULE_GUID);
         when(mockStudyDao.getStudy(TEST_APP_ID, TEST_STUDY_ID)).thenReturn(study);
 
+        Assessment assessment = new Assessment();
+        assessment.setGuid(ASSESSMENT_1_GUID);
+        assessment.setPhase(AssessmentPhase.REVIEW);
```
Comment (Erin-Mounts) [inline src/test/java/org/sagebionetworks/bridge/services/StudyServiceTest.java:784]: Is there a test somewhere in this file that checks that the assessment phase has actually changed to PUBLISHED after the study has moved to recruitment phase? Or maybe somewhere else? I don't remember seeing that anywhere but I might have just missed it.

**Review by Erin-Mounts — APPROVED** [review]
Couple of questions about tests. Code per se looks good to me.

File: src/test/java/org/sagebionetworks/bridge/services/StudyServiceTest.java (hunk 2 above)
Comment (nategbrown9) [reply src/test/java/org/sagebionetworks/bridge/services/StudyServiceTest.java:784]: Just added that check to this test.

File: src/main/java/org/sagebionetworks/bridge/validators/AssessmentValidator.java [hunk 3]
```diff
@@ -50,6 +46,9 @@ public void validate(Object target, Errors errors) {
             errors.rejectValue("title", CANNOT_BE_BLANK);   
         }
         validateStringLength(errors, 255, assessment.getTitle(), "title");
+        if (assessment.getPhase() == null) {
+            errors.rejectValue("phase", CANNOT_BE_NULL);
+        }
```
Comment (syoung-smallwisdom) [inline src/main/java/org/sagebionetworks/bridge/validators/AssessmentValidator.java:56]: What happens to existing assessments?

File: src/main/resources/db/changelog/changelog.sql [hunk 4]
```diff
@@ -1098,3 +1098,7 @@ ALTER TABLE `AdherenceRecords`
 ADD COLUMN `postProcessingAttributes` text COLLATE utf8_unicode_ci DEFAULT NULL,
 ADD COLUMN `postProcessingCompletedOn` bigint(20) unsigned DEFAULT NULL,
 ADD COLUMN `postProcessingStatus` varchar(255) DEFAULT NULL;
+
+-- changeset bridge:81
+ALTER TABLE `Assessments`
+ADD COLUMN `phase` enum('DRAFT', 'REVIEW', 'PUBLISHED') DEFAULT 'DRAFT',
```
Comment (syoung-smallwisdom) [inline src/main/resources/db/changelog/changelog.sql]: Shared assessments are published by default, aren't they?

File: src/test/java/org/sagebionetworks/bridge/models/assessments/AssessmentTest.java [hunk 5]
```diff
@@ -133,6 +133,7 @@ public void serializeRountrip() throws Exception {
         assertEquals(node.get("additionalMetadata").get("key1").textValue(), "value1");
         assertEquals(node.get("additionalMetadata").get("key2").textValue(), "value2");
         assertEquals(node.get("type").textValue(), "Assessment");
+        assertEquals(node.get("phase").textValue(), "draft");
```
Comment (syoung-smallwisdom) [inline src/test/java/org/sagebionetworks/bridge/models/assessments/AssessmentTest.java:136]: I'm a bit confused by how this works. Isn't the enum defined uppercase?

**Review by syoung-smallwisdom — APPROVED** [review]
Couple of questions. Otherwise, looks good.

File: src/main/java/org/sagebionetworks/bridge/validators/AssessmentValidator.java (hunk 3 above)
Comment (nategbrown9) [reply src/main/java/org/sagebionetworks/bridge/validators/AssessmentValidator.java:56]: The sql add column migration will set them all to Draft.

File: src/main/resources/db/changelog/changelog.sql (hunk 4 above)
Comment (nategbrown9) [reply src/main/resources/db/changelog/changelog.sql]: Good catch. Just added code to set Shared assessments to published when they are created. Not sure the best approach for migrating existing shared assessments.

File: src/test/java/org/sagebionetworks/bridge/models/assessments/AssessmentTest.java (hunk 5 above)
Comment (nategbrown9) [reply src/test/java/org/sagebionetworks/bridge/models/assessments/AssessmentTest.java:136]: Yes, but the json serializer is set to do enums as lowercase: https://github.com/Sage-Bionetworks/BridgeServer2/blob/4c7433da919f08c6be05b2e41f4141ebe9850a9d/src/main/java/org/sagebionetworks/bridge/json/BridgeObjectMapper.java#L77

Comment (syoung-smallwisdom) [reply src/test/java/org/sagebionetworks/bridge/models/assessments/AssessmentTest.java:136]: That's handy

File: src/main/java/org/sagebionetworks/bridge/models/assessments/HibernateAssessment.java [hunk 6]
```diff
@@ -78,6 +78,7 @@ public static HibernateAssessment create(String appId, Assessment dto) {
         assessment.setMinAge(dto.getMinAge());
         assessment.setMaxAge(dto.getMaxAge());
         assessment.setAdditionalMetadata(dto.getAdditionalMetadata());
+        assessment.setPhase(dto.getPhase());
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/models/assessments/HibernateAssessment.java:83]: Do we need to update HibernateAssessmentTest?

File: src/main/java/org/sagebionetworks/bridge/validators/AssessmentValidator.java [hunk 7]
```diff
@@ -2,11 +2,7 @@
 
 import static org.apache.commons.lang3.StringUtils.isBlank;
 import static org.sagebionetworks.bridge.BridgeConstants.SHARED_APP_ID;
-import static org.sagebionetworks.bridge.validators.Validate.BRIDGE_EVENT_ID_ERROR;
-import static org.sagebionetworks.bridge.validators.Validate.BRIDGE_EVENT_ID_PATTERN;
-import static org.sagebionetworks.bridge.validators.Validate.CANNOT_BE_BLANK;
-import static org.sagebionetworks.bridge.validators.Validate.CANNOT_BE_NEGATIVE;
-import static org.sagebionetworks.bridge.validators.Validate.INVALID_TYPE;
+import static org.sagebionetworks.bridge.validators.Validate.*;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/validators/AssessmentValidator.java]: The code convention for Bridge Server is that we don't use wild cards for imports. My IntelliJ config is import with * and static import with * is set to 999 (basically never). The sort order is

* import static
* (blank line)
* import java.*
* import javax.*
* (blank line)
* all other imports
* (blank line)
* import org.sagebionetworks.bridge.*

I can send you my code style XML if that's easier.

File: src/main/java/org/sagebionetworks/bridge/validators/AssessmentValidator.java (hunk 3 above)
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/validators/AssessmentValidator.java:56]: Do we need to add a test case to AssessmentValidatorTest?

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
A few comments about tests and imports. Otherwise, this looks good.

**Review by DwayneJengSage — APPROVED** [review]
Looks good. Just fix that one EnumeratedType annotation we discovered, and it should be good.

**Review by DwayneJengSage — APPROVED** [review]

**Pull request merged**